File: tiny-render-py/algebra_py/vec2.py


from array import array

from .common import *
from math import sqrt

class vec2(object):
    __slots__ = ["data"]

    def __init__(self, x: float=0.0, y: float=0.0) -> None:
        self.data = array("f", [x, y])

    def __repr__(self) -> str:
        return """
{:.4f} {:.4f}
""".format(self.data[0],  self.data[1])

    def __setitem__(self, key: int, value: float):
        self.data[key] = value

    def __getitem__(self, key: int) -> float:
        return self.data[key]

    def tobytes(self):
        return self.data.tobytes()
        
def vec2Lenght(v: vec2) -> float:
	return sqrt(v[0]*v[0] +
		        v[1]*v[1])

def vec2Normalize(v: vec2) -> vec2:
    len = vec2Lenght(v)

    rt = vec2()

    if len != 0.0: 
        rt[0] = v[0] / len
        rt[1] = v[1] / len
	
    return rt

def vec2Scale(v: vec2, scale: float) -> vec2:
    rt = v

    rt[0] *= scale
    rt[1] *= scale

    return rt

def vec2Invert(v: vec2) -> vec2:
    rt = vec2()

    rt[0] = -v[0]
    rt[1] = -v[1]

    return rt

def vec2Dot(a: vec2, b: vec2) -> float:
	return a[0]*b[0] + a[1]*b[1] + a[2]*b[2]

def vec2Sum(a: vec2, b: vec2) -> vec2:
    rt = vec2()

    rt[0] = a[0] + b[0]
    rt[1] = a[1] + b[1]

    return rt

def vec2Sub(a: vec2, b: vec2) -> vec2:
    rt = vec2()

    rt[0] = a[0] - b[0]
    rt[1] = a[1] - b[1]

    return rt

File: LANG-jsts/tiny-render/app/algebra/vec2.js
import { fEPS } from "./common.js";
export class vec2 {
    data;
    order = 2;
    constructor(x, y) {
        if (x instanceof vec2) {
            this.data = new Float32Array(2);
            this.fromVec2(x);
        }
        else {
            this.data = new Float32Array(2);
            this.data[0] = x || 0.0;
            this.data[1] = y || 0.0;
        }
    }
    fromVec2(src) {
        this.data[0] = src.data[0];
        this.data[1] = src.data[1];
    }
    lenght() {
        return Math.sqrt(this.data[0] * this.data[0] +
            this.data[1] * this.data[1]);
    }
    normalize() {
        let len;
        len = this.lenght();
        if (len != 0.0) {
            this.data[0] = this.data[0] / len;
            this.data[1] = this.data[1] / len;
        }
    }
    scale(scale) {
        this.data[0] = this.data[0] * scale;
        this.data[1] = this.data[1] * scale;
    }
    invert() {
        this.data[0] = -this.data[0];
        this.data[1] = -this.data[1];
    }
}
;
export function vec2Normalize(v) {
    let rt = new vec2();
    let len;
    len = v.lenght();
    if (len > fEPS) {
        rt.data[0] = v.data[0] / len;
        rt.data[1] = v.data[1] / len;
    }
    return rt;
}
function vec2Scale(v, scale) {
    let rt = new vec2();
    rt.data[0] = v.data[0] * scale;
    rt.data[1] = v.data[1] * scale;
    return rt;
}
function vec2Invert(v) {
    let rt = new vec2;
    rt.data[0] = -v.data[0];
    rt.data[1] = -v.data[1];
    return rt;
}
function vec2Dot(a, b) {
    return a.data[0] * b.data[0] + a.data[1] * b.data[1] + a.data[2] * b.data[2];
}
function vec2Sum(a, b) {
    let rt = new vec2;
    rt.data[0] = a.data[0] + b.data[0];
    rt.data[1] = a.data[1] + b.data[1];
    return rt;
}
function vec2Sub(a, b) {
    let rt = new vec2;
    rt.data[0] = a.data[0] - b.data[0];
    rt.data[1] = a.data[1] - b.data[1];
    return rt;
}
